### PR TITLE
12-food-distribution

### PR DESCRIPTION
**Pull request by Urhengulas**
https://trello.com/c/fx2RxXOz/12-food-distribution

 - 4-dim mechanical wave function

### CODE REVIEW

**Comment by Urhengulas** [comment]
[Food distribution](https://trello.com/c/fx2RxXOz/12-food-distribution)

File: food.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    # general energy consumption
+    cell.energy_level -= conf.GENERAL_ENERGY_CONSUMPTION
+
+    # other energ consumption (in future)
+    cell.energy_level -= 0
+
+    return cell
+
+
+def alive(cell):
```
Comment (Animuslupus) [inline food.py]: `is_alive` is already implemented in cis_cell.py, maybe you should import it

**Review by elli-wolf — CHANGES_REQUESTED** [review]
https://github.com/codeuniversity/al-cis/blob/479cc7f1db9f6c72883eaa5f638426e163e843a3/food.py#L18-L22
This loop could be in `cell_computing_service` to enhance the readability of the `cell_computing_service`.

**Review by elli-wolf — CHANGES_REQUESTED** [review]
You should maybe think about, replacing the function generators with actual functions. Since you don't gain any benefit through the use of function generators.

**Review by elli-wolf — CHANGES_REQUESTED** [review]
I think it would be useful if you would integrate these changes into `cis_env`.

File: cell_computing_service.py [hunk 2]
```diff
@@ -32,17 +36,16 @@ def ComputeCellInteractions(self, incoming_batch, context):
...
         for c in incoming_batch.cells_to_compute:
-            cis_env.feed_cell(
-                c,
-                id_to_cell,
-                id_to_cell_energy_averaged,
-                food_factor=conf.WANTED_CELL_AMOUNT_PER_BUCKET /
-                len(
-                    incoming_batch.cells_to_compute))
+            c = cis_env.feed(c, incoming_batch.time_step)
+
```
Comment (alexmorten) [inline cell_computing_service.py]: FYI: After feeding all the cells, it is probably a good idea to still average the energy level inside a connected cell group with `average_out_energy_in_connected_cells` (you removed that call in the merge)

File: cis_env.py [hunk 3]
```diff
@@ -14,18 +15,56 @@ def curl(pos, vec_len=0.004, coor_origin=conf.CURL_CENTRE):
...
+    food_value = normalize(food_value)
+
+    # check if cell gets (the fixed amount of) food
+    if food_value < conf.FOOD_THRESHOLD:
+        cell.energy_level += conf.FOOD_ENERGY
+
+    return cell
+
+
+def normalize(num, definition_area_size=6):
```
Comment (Animuslupus) [inline cis_env.py]: The function name `normalize` is too brought, normalization can reference to wide range of procedures and processes, it isn't clear which one of them you reference.

File: cis_env.py [hunk 4]
```diff
@@ -14,18 +15,56 @@ def curl(pos, vec_len=0.004, coor_origin=conf.CURL_CENTRE):
...
+    """
+        Normalizes the defenition area of the num to {0, 1} .
```
Comment (Animuslupus) [inline cis_env.py]: There is a spelling mistake in definition :sweat_smile:

File: cis_env.py [hunk 5]
```diff
@@ -3,6 +3,7 @@
 import numpy as np
 import dna_decoding
 from cis_cell import random_vector_of_length
+import scipy.constants as cons
```
Comment (Animuslupus) [inline cis_env.py]: If you want to use scipy in CIS you have to add it to the dependencies. But maybe you also find a more elegant way to import the speed of light :wink:.

**Review by elli-wolf — CHANGES_REQUESTED** [review]

**Pull request merged**